processor/src/advice/mem_provider.rs

```diff
@@ -0,0 +1,174 @@
+use super::{
+    AdviceProvider, AdviceSet, BTreeMap, ExecutionError, Felt, IntoBytes, ProgramInputs,
+    StarkField, Vec, Word,
+};
+
+// MEMORY ADVICE PROVIDER
+// ================================================================================================
+
+/// An in-memory `[AdviceProvider]` implementation to support program execution.
+///
+/// Uses `[BTreeMap]` as backend.
+#[derive(Debug, Clone, Default)]
+pub struct MemAdviceProvider {
+    step: u32,
+    tape: Vec<Felt>,
+    values: BTreeMap<[u8; 32], Vec<Felt>>,
+    sets: BTreeMap<[u8; 32], AdviceSet>,
+}
+
+#[cfg(test)]
+impl MemAdviceProvider {
+    // ADVISE SETS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns true if the advice set with the specified root is present in this advice provider.
+    pub fn has_advice_set(&self, root: Word) -> bool {
+        self.sets.contains_key(&root.into_bytes())
+    }
+}
```

Comment:
minor nit: I'd probably move this to the end of the file and get rid of the `ADVICE SETS` separator.

Reply:
True, should be there indeed. Fixed!